kernel/kernel-api/src/main/java/io/delta/kernel/internal/metrics/SnapshotQueryContext.java

```diff
@@ -32,35 +35,46 @@ public class SnapshotQueryContext {
 
   /** Creates a {@link SnapshotQueryContext} for a Snapshot created by a latest snapshot query */
   public static SnapshotQueryContext forLatestSnapshot(String tablePath) {
-    return new SnapshotQueryContext(tablePath, Optional.empty(), Optional.empty());
+    return new SnapshotQueryContext(
+        tablePath, Optional.empty(), Optional.empty(), Optional.empty());
   }
 
   /** Creates a {@link SnapshotQueryContext} for a Snapshot created by a AS OF VERSION query */
   public static SnapshotQueryContext forVersionSnapshot(String tablePath, long version) {
-    return new SnapshotQueryContext(tablePath, Optional.of(version), Optional.empty());
+    return new SnapshotQueryContext(
+        tablePath, Optional.of(version), Optional.empty(), Optional.empty());
   }
 
   /** Creates a {@link SnapshotQueryContext} for a Snapshot created by a AS OF TIMESTAMP query */
   public static SnapshotQueryContext forTimestampSnapshot(String tablePath, long timestamp) {
-    return new SnapshotQueryContext(tablePath, Optional.empty(), Optional.of(timestamp));
+    return new SnapshotQueryContext(
+        tablePath, Optional.empty(), Optional.empty(), Optional.of(timestamp));
   }
 
   private final String tablePath;
-  private Optional<Long> version;
   private final Optional<Long> providedTimestamp;
   private final SnapshotMetrics snapshotMetrics = new SnapshotMetrics();
 
+  private Optional<Long> version;
+  private Optional<Long> checkpointVersion;
+
   /**
    * @param tablePath the table path for the table being queried
    * @param providedVersion the provided version for a time-travel-by-version query, empty if this
    *     is not a time-travel-by-version query
+   * @param checkpointVersion the version of the checkpoint used for this snapshot, empty if no
+   *     checkpoint was used or if this is a failed snapshot construction
    * @param providedTimestamp the provided timestamp for a time-travel-by-timestamp query, empty if
    *     this is not a time-travel-by-timestamp query
    */
   private SnapshotQueryContext(
-      String tablePath, Optional<Long> providedVersion, Optional<Long> providedTimestamp) {
+      String tablePath,
+      Optional<Long> providedVersion,
+      Optional<Long> checkpointVersion,
```

Comment:
nit - but if this is never provided why not omit it from the arg list and just initialize it as empty?